doc/ldapdebug.1.in
.\"
.\" LDAP Utilities
.\" Copyright (C) 2019 David M. Syzdek <david@syzdek.net>.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"    1. Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"    2. Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in the
.\"       documentation and/or other materials provided with the distribution.
.\"
.\"    3. Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived from
.\"       this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
.\" IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
.\" THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" doc/ldapdebug.1.in - man page for ldapdebug
.\"
.TH "LDAPDEBUG" "1" "@RELEASE_MONTH@" "@PACKAGE_NAME@" "User Commands"
.SH NAME
ldapdebug \- LDAP displays values from ldap_get_option(3).


.SH SYNOPSIS
\fBldapdebug\fR
[\fB-b\fR \fIbasedn\fR]
[\fB-c\fR]
[\fB-d\fR \fIlevel\fR]
[\fB-D\fR \fIbinddn\fR]
[\fB-H\fR \fIURI\fR]
[\fB-l\fR \fIlimit\fR]
[\fB-L\fR[\fB-L\fR]]
[\fB-n\fR]
[\fB-v\fR | \fB--version\fR]
[\fB-s\fR \fIscope\fR]
[\fB-S\fR \fIattr\fR]
[\fB-w\fR \fIpasswd\fR]
[\fB-W\fR]
[\fB-x\fR]
[\fB-y\fR \fIfile\fR]
[\fB-Y\fR \fImech\fR]
[\fB-z\fR \fIlimit\fR]
[\fB-Z\fR[\fB-Z\fR]]
.sp
\fBldapdebug\fR [ \fB-h\fR | \fB--help\fR ]
.sp
\fBldapdebug\fR [ \fB-V\fR | \fB--version\fR ]

.SH DESCRIPTION
\fBldapdebug\fR is a utility which initiates a connection the an LDAP server
and optionally binds to the LDAP server.  Once the LDAP session has been
established, the utility will display the values of various options
available from \fBldap_get_option\fR(3).  This utility is useful for debugging
\fIldap.conf\fR and \fI.ldaprc\fR files.

.SH OPTIONS
.TP
\fB-c\fR
do not stop if an error is encountered
.TP
\fB-d\fR
set OpenLDAP debug level to `level'
.TP
\fB-D\fR \fIbinddn\fR
bind DN used for simple bind
.TP
\fB-H\fR \fIURI\fR
specifies list of LDAP Uniform Resource Identifier(s) used to connect to LDAP server
.TP
\fB-l\fR \fIlimit\fR
time limit (in seconds) for search
.TP
\fB-L\fR
two \fB-L\fR disables comments
.TP
\fB-v\fR   \fB--version\fR
run in verbose mode
.TP
\fB-s\fR \fIscope\fR
specifies search filter. Must be one of \fIbase\fR, \fIone\fR, \fIsub\fR, or \fIchild\fR
.TP
\fB-S\fR \fIattr\fR
sort results by attribute \fIattr\fR
.TP
\fB-w\fR \fIpasswd\fR
bind password used for simple bind
.TP
\fB-W\fR
prompt for bind password used in simple bind
.TP
\fB-x\fR
use simple authentication for bind
.TP
\fB-y\fR \fIfile\fR
read bind password from file
.TP
\fB-Y\fR \fImech\fR
SASL mechanism used during bind
.TP
\fB-z\fR \fIlimit\fR
size limit for search
.TP
\fB-Z\fR[\fB-Z\fR]
Issue  StartTLS before bind request. \fB-ZZ\fR requires TLS operations to be successful.


.SH "SEE ALSO"
.BR ldapsearch (1),
.BR ldap2csv (1),
.BR ldaptree (1),
.BR ldap_get_option (3),
.BR ldap.conf (5)


.SH AUTHOR
David M. Syzdek <david@syzdek.net>


.SH ACKNOWLEDGEMENTS
\fB@PACKAGE_NAME@\fR is developed and maintained by David M. Syzdek
<david@syzdek.net>. \fB@PACKAGE_NAME@\fR utilizes and is styled after the
tools and libraries maintained by the \fBThe OpenLDAP Project\fR
<http://www.openldap.org/>.


.Sh CAVEATS

.\" end of man page


ChangeLog

LDAP Utilities
Copyright (C) 2012 Bindle Binaries <syzdek@bindlebinaries.com>.

0.3
      - refactoring libldaputils (syzdek)
      - adding TLS support

0.2   
      - Recreating Xcode project file. (syzdek)
      - Removing files installed by autotools from git repo. (syzdek)
      - Updating autotools files (Makefile.am, configure.ac, etc). (syzdek)
      - Changing versioning scheme. (syzdek)
      - Re-writing README. (syzdek)

0.1.1
      - Updating ldaputils_getpass() to use stderr.
      - Modifying behavior of --verbose to set debug level
      - Removing debugging lines for LDAP API Version.
      - Updating version information to include LDAP API information.
      - Adding sorting functions for LDAP searches.
      - Fixing problems compiling with MinGW.
      - Abstracting code for separating multivalue values.
      - Fixing gettext international support.
      - Changing how LDAP host/port/URI values are stored so that values
        from ldap.conf and ldaprc files will be used if they are not
        specified on the command line. (syzdek)
      - Changing default search scope to SUBTREE. (syzdek)
      - Changing version major version from 1.0.0. to 0.1.0 (syzdek)

0.1.0   2008/10/12
	* initial release of package (syzdek)


src/ldapdebug.c
/*
 *  LDAP Utilities
 *  Copyright (C) 2012, 2019 David M. Syzdek <david@syzdek.net>.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are
 *  met:
 *
 *     1. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *
 *     2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *
 *     3. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 *  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 *  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/**
 *  @file src/ldapdebug.c export LDAP data to CSV file
 */
/*
 *  Simple Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldapdebug" -Wall -I../include'
 *     gcc ${CFLAGS} -c ldapdebug.c
 *     gcc ${CFLAGS} -c ldaputils_config.c
 *     gcc ${CFLAGS} -c ldaputils_config_opts.c
 *     gcc ${CFLAGS} -o ldapdebug ldapdebug.o ldaputils_config.o \
 *             ldaputils_config_opts.o -lldap
 *
 *  Libtool Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldapdebug" -Wall -I../include'
 *     libtool --mode=compile --tag=CC gcc ${CFLAGS} -c ldapdebug.c
 *     libtool --mode=compile --tag=CC gcc ${CFLAGS} -c ldaputils_config.c
 *     libtool --mode=compile --tag=CC gcc ${CFLAGS} -c ldaputils_config_opts.c
 *     libtool --mode=link    --tag=CC gcc ${CFLAGS} -o ldapdebug \
 *             ldapdebug.o ldaputils_config.o ldaputils_config_opts.o \
 *             -lldap
 *
 *  Libtool Clean:
 *     libtool --mode=clean rm -f ldapdebug.lo ldaputils_config.lo \
 *             ldaputils_config_opts.lo ldapdebug
 */
#define _LDAP_UTILS_SRC_LDAPDEBUG 1
#undef __LDAPUTILS_PMARK


///////////////
//           //
//  Headers  //
//           //
///////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Headers
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <getopt.h>

#define LDAP_DEPRECATED 1
#include <ldap.h>
#include <ldaputils.h>


///////////////////
//               //
//  Definitions  //
//               //
///////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Definitions
#endif

#ifndef PROGRAM_NAME
#define PROGRAM_NAME "ldapdebug"
#endif

#define MY_SHORT_OPTIONS LDAPUTILS_OPTIONS_COMMON LDAPUTILS_OPTIONS_SEARCH "o:"


//////////////////
//              //
//  Prototypes  //
//              //
//////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Prototypes
#endif

// main statement
int main(int argc, char * argv[]);

// parses configuration
int my_config(int argc, char * argv[], LDAPUtils ** ludp);


/////////////////
//             //
//  Functions  //
//             //
/////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Functions
#endif

/// prints program usage and exits
void ldaputils_usage(void)
{
   printf("Usage: %s [options] [filter [attributes...]]\n", PROGRAM_NAME);
   ldaputils_usage_search(MY_SHORT_OPTIONS);
   ldaputils_usage_common(MY_SHORT_OPTIONS);
   printf("\nReport bugs to <%s>.\n", PACKAGE_BUGREPORT);
   return;
}


/// main statement
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
int main(int argc, char * argv[])
{
   LDAPUtils * lud;

   if ((my_config(argc, argv, &lud)))
      return(1);
   if (!(lud))
      return(0);

   ldaputils_bind_s(lud);

   ldaputils_params(lud);

   ldaputils_unbind(lud);

   return(0);
}


/// parses configuration
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
int my_config(int argc, char * argv[], LDAPUtils ** ludp)
{
   int               c;
   int               err;
   int               option_index;

   static char   short_options[] = MY_SHORT_OPTIONS;
   static struct option long_options[] =
   {
      {"help",          no_argument, 0, 'h'},
      {"verbose",       no_argument, 0, 'v'},
      {"version",       no_argument, 0, 'V'},
      {NULL,            0,           0, 0  }
   };

   option_index = 0;
   *ludp        = NULL;

   // initialize ldap utilities
   if ((err = ldaputils_initialize(ludp, PROGRAM_NAME)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldaputils_initialize(): %s\n", PROGRAM_NAME, ldap_err2string(err));
      ldaputils_unbind(*ludp);
      return(1);
   };

   // loops through args
   while((c = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1)
   {
      switch(ldaputils_getopt(*ludp, c, optarg))
      {
         // shared option exit without error
         case -2:
         ldaputils_unbind(*ludp);
         *ludp = NULL;
         return(0);

         // no more arguments
         case -1:
         break;

         // long options toggles
         case 0:
         break;

         // shared option error
         case 1:
         ldaputils_unbind(*ludp);
         *ludp = NULL;
         return(1);

         // argument error
         case '?':
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         ldaputils_unbind(*ludp);
         *ludp = NULL;
         return(1);

         default:
         fprintf(stderr, "%s: unrecognized option `--%c'\n", PROGRAM_NAME, c);
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         ldaputils_unbind(*ludp);
         *ludp = NULL;
         return(1);
      };
   };

   // saves filter
   if (argc > optind)
      (*ludp)->filter = argv[optind];

   if (argc <= (optind+1))
      return(0);

   // configures LDAP attributes to return in results
   if (!((*ludp)->attrs = (char **) malloc(sizeof(char *) * (size_t)(argc-optind))))
   {
      fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
      ldaputils_unbind(*ludp);
      return(1);
   };
   for(c = 0; c < (argc-optind-1); c++)
      (*ludp)->attrs[c] = argv[optind+1+c];
   (*ludp)->attrs[c] = NULL;

   // reads password
   if ((err = ldaputils_pass(*ludp)) != 0)
   {
      ldaputils_unbind(*ludp);
      return(1);
   };

   return(0);
}

/* end of source file */
